src/lib/logging/Logger.js

```diff
@@ -0,0 +1,314 @@
+     * @return {Function} A function that can be invoked with a string, to log a message.
+     */
+    getLoggerFunction(code) {
+        let logFunction;
+
+        switch (code) {
+            case LOG_CODES.warning:
+                logFunction = this.logger.warn;
+                break;
+            case LOG_CODES.error:
+                logFunction = this.logger.error;
+                break;
+            case LOG_CODES.info:
+                logFunction = this.logger.info;
+                break;
+            default:
+                logFunction = null;
+        }
+
+        return logFunction;
+    }
+
+    /**
+     * Handler for uncaught errors. Logs and caches
+     *
+     * @private
+     * @param {Error} error - Error object provided by the error event.
+     * @return {void}
+     */
+    onUncaughtError(error) {
+        this.error(error.message, '\n', error.error.stack);
+    }
+
+    /**
+     * Commit a message to the cache and run the appropriate log function.
+     *
+     * @private
+     * @param {LOG_CODES|string} code - The log code this belongs to.
+     * @param {*} message - The message to log.
+     * @return {void}
+     */
+    commitMessage(code, message) {
+        const logFunction = this.getLoggerFunction(code);
+
+        // Format message and add a timestamp
+        const timestamp = new Date().toISOString();
+        this.cache.add(code, timestamp, message);
+
+        // Also wrapping the code into the message
+        // #TODO(@jholdstock): abstract this step
+        if (logFunction) {
+            const formattedMessage = `${timestamp} "${message}"`;
+            logFunction(`[${code}] ${formattedMessage}`);
+        }
+    }
+
+    /**
+     * Log an info message to the logger, and commit to the cache.
+     *
+     * @public
+     * @param {*} args - Arguments to be printed to the log and cache.
+     * @return {void}
+     */
+    info(...args) {
+        const message = arrayToString(args);
+        this.commitMessage(LOG_CODES.info, message);
+    }
+
+    /**
+     * Log a warning message to the logger, and commit to the cache.
+     *
+     * @public
+     * @param {*} args - Arguments to be printed to the log and cache.
+     * @return {void}
+     */
+    warn(...args) {
+        const message = arrayToString(args);
+        this.commitMessage(LOG_CODES.warning, message);
+    }
+
+    /**
+     * Log an error message to the logger, and commit to the cache.
+     *
+     * @public
+     * @param {*} args - Arguments to be printed to the log and cache.
+     * @return {void}
+     */
+    error(...args) {
+        const message = arrayToString(args);
+        this.commitMessage(LOG_CODES.error, message);
+    }
+
+    /**
+     * Log a metric message to the logger, and commit to the cache.
+     *
+     * @public
+     * @param {number} code - Code associated with a specific metric.
+     * @param {*} value - 
+     * @return {void}
+     */
+    metric(code, value) {
```

Comment:
Would it be potentially feasible to use this method as a replacement for fileMetrics? They're both used for the same purpose of saving metrics.